File: GraalVM_InstallerForWindows/Core/Installer.cs
/*
Copyright (C) 2021  Nikos Siatras
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.
SourceRabbit GCode Sender is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
You should have received a copy of the GNU General Public License
along with this program.
*/

using GraalVM_InstallerForWindows.Core;
using ICSharpCode.SharpZipLib.Core;
using ICSharpCode.SharpZipLib.Zip;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Net;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace GraalVM_InstallerForWindows
{
    public class Installer
    {
        private string fStatus = "";
        private int fProgressValue = 0;
        private string fVersionToInstall, fInstallationFolder;

        private Thread fInstallationThread;
        private ManualResetEvent fWaitStepToFinish = new ManualResetEvent(false);
        private bool fFinished = false;


        public Installer()
        {

        }


        public void InstallGraalVM(string versionName, string installationPath)
        {
            fVersionToInstall = versionName;
            fInstallationFolder = installationPath;

            fInstallationThread = new Thread(InstallationThreadStart);
            fInstallationThread.Start();
        }


        private void InstallationThreadStart()
        {
            fWaitStepToFinish.Reset();
            Step1_DownloadGraalVM();
            fWaitStepToFinish.WaitOne();

            fWaitStepToFinish.Reset();
            Step2_UnzipGraalVM();
            fWaitStepToFinish.WaitOne();


            fWaitStepToFinish.Reset();
            Step3_SetEnvironmentalVariables();
            fWaitStepToFinish.WaitOne();

            fWaitStepToFinish.Reset();
            Step4_RunJarFix();
            fWaitStepToFinish.WaitOne();


            // Set fFinished to true at the end!!!
            fFinished = true;
        }


        private void Step1_DownloadGraalVM()
        {
            fStatus = "Downloading " + fVersionToInstall + "...";

            string webPathToGraalVMZip = GraalVMVersionManager.GraalVMVersions[fVersionToInstall];
            string downloadPathToDisk = fInstallationFolder + "\\" + fVersionToInstall;

            // Create installation folder if doesn't exit 
            if(!Directory.Exists(fInstallationFolder))
            {
                Directory.CreateDirectory(fInstallationFolder);
            }

            if (!File.Exists(downloadPathToDisk))
            {
                WebClient webClient = new WebClient();
                webClient.DownloadProgressChanged += new DownloadProgressChangedEventHandler(Event_DownloadProgressChanged);
                webClient.DownloadFileCompleted += new AsyncCompletedEventHandler(Event_DownloadFileFinished);
                webClient.DownloadFileAsync(new Uri(webPathToGraalVMZip), downloadPathToDisk);
            }
            else
            {
                fProgressValue = 100;
                fWaitStepToFinish.Set();
            }
        }

        private void Event_DownloadProgressChanged(object sender, DownloadProgressChangedEventArgs e)
        {
            fProgressValue = e.ProgressPercentage;
        }

        private void Event_DownloadFileFinished(object sender, AsyncCompletedEventArgs e)
        {
            fWaitStepToFinish.Set();
        }


        private void Step2_UnzipGraalVM()
        {
            fProgressValue = 0;
            fStatus = "Extracting GraalVM " + fVersionToInstall + ". Please wait...";
            string zipPath = fInstallationFolder + "\\" + fVersionToInstall;

            Stream zipStream = new FileStream(zipPath, FileMode.Open);

            string outFolder = fInstallationFolder;
            outFolder = outFolder.EndsWith("\\") ? outFolder.Substring(0, outFolder.Length - 1) : outFolder;


            // Get how many files needs to be extracted
            int totalFilesToExtract = 0;
            using (var zipInputStream = new ZipInputStream(zipStream))
            {
                while (zipInputStream.GetNextEntry() is ZipEntry zipEntry)
                {
                    totalFilesToExtract += 1;
                }
            }

            // Extract files
            zipStream = new FileStream(zipPath, FileMode.Open);
            int filesExtracted = 0;
            using (var zipInputStream = new ZipInputStream(zipStream))
            {

                while (zipInputStream.GetNextEntry() is ZipEntry zipEntry)
                {
                    string entryFileName = zipEntry.Name;
                    entryFileName = entryFileName.Substring(entryFileName.IndexOf("/"));
                    entryFileName = entryFileName.Replace("/", "\\");

                    // 4K is optimum
                    var buffer = new byte[4096];

                    // Manipulate the output filename here as desired.
                    var fullZipToPath = outFolder + entryFileName;
                    var directoryName = Path.GetDirectoryName(fullZipToPath);
                    if (directoryName.Length > 0)
                    {
                        Directory.CreateDirectory(directoryName);
                    }

                    // Skip directory entry
                    if (Path.GetFileName(fullZipToPath).Length == 0)
                    {
                        continue;
                    }

                    using (FileStream streamWriter = File.Create(fullZipToPath))
                    {
                        StreamUtils.Copy(zipInputStream, streamWriter, buffer);
                    }

                    filesExtracted += 1;

                    fProgressValue = (int)(filesExtracted * 100) / totalFilesToExtract;
                }
            }

            fProgressValue = 100;
            fWaitStepToFinish.Set();
        }


        private void Step3_SetEnvironmentalVariables()
        {
            fStatus = "Setting environmental variables...";

            // Create the JAVA_HOME environmental variable
            string environmentalVariableName = "JAVA_HOME";
            Environment.SetEnvironmentVariable(environmentalVariableName, fInstallationFolder, EnvironmentVariableTarget.Machine);

            fProgressValue = 50;

            // Create Path
            // Example C:\GraalVM\bin
            string pathVar = fInstallationFolder + "\\bin\\";
            var name = "PATH";
            var scope = EnvironmentVariableTarget.Machine;
            var oldValue = Environment.GetEnvironmentVariable(name, scope);
            if (!oldValue.Contains(pathVar))
            {
                if (!oldValue.EndsWith(";"))
                {
                    oldValue = oldValue + ";";
                }
                var newValue = oldValue + pathVar + ";";
                Environment.SetEnvironmentVariable(name, newValue, scope);
            }

            fProgressValue = 100;
            fWaitStepToFinish.Set();
        }

        private void Step4_RunJarFix()
        {
            string appPath = AppDomain.CurrentDomain.BaseDirectory;
            string jarFixPath = appPath + "JarFix\\jarfix.exe";
            System.Diagnostics.Process.Start(jarFixPath);
            fWaitStepToFinish.Set();
        }


        public string Status
        {
            get { return fStatus; }
        }

        public int Progress
        {
            get { return fProgressValue; }
        }

        public bool IsFinished
        {
            get { return fFinished; }
        }
    }
}

File: NetBeans Project/src/graalvminstallerforwindows/Core/Installer.java
package graalvminstallerforwindows.Core;

import graalvminstallerforwindows.Core.Utilities.DosPromt;
import graalvminstallerforwindows.Core.Utilities.EnvironmentVariablesManager;
import graalvminstallerforwindows.Core.Utilities.FileUtils;
import graalvminstallerforwindows.Core.Utilities.Unzipper;
import graalvminstallerforwindows.UI.UITools;
import graalvminstallerforwindows.UI.frmDownloadFile;
import graalvminstallerforwindows.UI.frmMain;
import java.io.File;
import java.io.FilenameFilter;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;

/*
    Copyright (C) 2024 Nikolaos Siatras
    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.
    This software is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
    You should have received a copy of the GNU General Public License
    along with this program.
 */
/**
 *
 * @author Nikos Siatras
 */
public class Installer
{

    private final String fDownloadURL, fInstallationPath;
    private final String fDownloadedZipFilePath;
    private final FileUtils fFileUtils;
    private boolean fEnvironmentVariablesAreSet = false;

    public Installer(String downloadURL, String installationPath)
    {
        fDownloadURL = downloadURL;
        fInstallationPath = installationPath;
        fDownloadedZipFilePath = fInstallationPath + "\\GraalVMDownload.zip";
        fFileUtils = new FileUtils();
    }

    /**
     * Returns true if installation finishes successfully
     *
     * @return
     * @throws Exception
     */
    public boolean IntallGrallVM() throws Exception
    {
        boolean stepFinished = false;
        STEP1_PrepareInstallationFolder();

        stepFinished = STEP2_DownloadGraalVM();
        if (!stepFinished)
        {
            return false;
        }

        stepFinished = STEP3_UnzipDownloadedFile();
        if (!stepFinished)
        {
            return false;
        }

        stepFinished = STEP4_MoveUnzippedGraalVMToInstallationPath();
        if (!stepFinished)
        {
            return false;
        }

        STEP5_SetJavaHomePathAndRunJarFix();

        return true;
    }

    private void STEP1_PrepareInstallationFolder() throws Exception
    {
        // Check if installation directory exists and it is empty!
        File installationDir = new File(fInstallationPath);

        if (!fFileUtils.CheckIfDirectoryExists(fInstallationPath))
        {
            // Create installationDir if not exists
            installationDir.mkdir();
        }

        if (!fFileUtils.CheckIfDirectoryExistsAndItIsEmpty(fInstallationPath))
        {
            throw new Exception("Directory '" + fInstallationPath + "' is not empty!\nDelete any files and directories inside the installation path and try again.");
        }
    }

    private boolean STEP2_DownloadGraalVM() throws Exception
    {
        boolean downloadFinishedSuccessfully = true;

        final String savePath = fDownloadedZipFilePath;

        frmDownloadFile frm = new frmDownloadFile(frmMain.fInstance, true, fDownloadURL, savePath);
        downloadFinishedSuccessfully = frm.OpenAndDownloadFile();

        return downloadFinishedSuccessfully;
    }

    private boolean STEP3_UnzipDownloadedFile() throws Exception
    {
        UITools.ShowPleaseWaitDialog("Please wait", "Extracting GraalVMDownload.zip<br>This can take a while...", frmMain.fInstance, () ->
        {
            try
            {
                Unzipper.Unzip(fDownloadedZipFilePath, fInstallationPath + "\\Unzipped");

                // Delete GraalVMDownload.zip
                File fileToDelete = new File(fDownloadedZipFilePath);
                fileToDelete.delete();
            }
            catch (Exception ex)
            {
            }
        });

        return true;
    }

    private boolean STEP4_MoveUnzippedGraalVMToInstallationPath() throws Exception
    {
        File installationDir = new File(fInstallationPath + "\\Unzipped");

        String[] directories = installationDir.list(new FilenameFilter()
        {
            @Override
            public boolean accept(File current, String name)
            {
                return new File(current, name).isDirectory();
            }
        });

        String sourceDir = fInstallationPath + "\\Unzipped\\" + directories[0];
        fFileUtils.MoveDirContentsToOtherDir(sourceDir, fInstallationPath);

        Files.delete(Paths.get(fInstallationPath + "\\Unzipped\\"));

        return true;
    }

    private boolean STEP5_SetJavaHomePathAndRunJarFix() throws Exception
    {

        UITools.ShowPleaseWaitDialog("Please wait", "Setting Environment variables...", frmMain.fInstance, () ->
        {
            try
            {
                EnvironmentVariablesManager.SetEnvironmentVariable("JAVA_HOME", fInstallationPath);
                EnvironmentVariablesManager.AddEnvironmentVariable("PATH", fInstallationPath + "\\bin\\");
                fEnvironmentVariablesAreSet = true;
            }
            catch (Exception ex)
            {
                fEnvironmentVariablesAreSet = false;
                return;
            }

            try
            {
                // Wait 2 seconds before run JarFix
                Thread.sleep(2000);
            }
            catch (Exception ex)
            {

            }
        });

        if (!fEnvironmentVariablesAreSet)
        {
            throw new Exception("Cannot set environment variables!");
        }

        // Call Jar Fix !!
        try
        {
            // Create a new  jarfix.ini
            String jarFixIniText = fFileUtils.ReadTextFile("Prerequisites/jarfix_template.ini");
            String newIniFile = jarFixIniText.replace("#PATH_TO_GRAALVM#", fInstallationPath);
            fFileUtils.SaveTextFile("Prerequisites/jarfix.ini", newIniFile);

            // Run jarfix.exe
            final String userDir = System.getProperty("user.dir");
            String jarFixRunCommand = userDir + "\\Prerequisites\\jarfix.exe";
            DosPromt.ExecuteDosPromtAndWaitToFinish(jarFixRunCommand);
        }
        catch (IOException ex)
        {
            throw new Exception("Could not run JAR Fix!");
        }

        return true;
    }

}
